### PR TITLE
Port link_rfu from Emerald

### PR DESCRIPTION
**Pull request by PikalaxALT**

### CODE REVIEW

File: src/link_rfu.c [hunk 1]
```diff
@@ -0,0 +1,1382 @@
...
+    case 48:
+        if (r6 == 0)
+        {
+            stwiRecvBuffer = rfu_getSTWIRecvBuffer()->rxPacketAlloc.rfuPacket8.data;
+            gUnknown_3005E10.unk_14 = stwiRecvBuffer[8];
+            sub_80FE83C(gUnknown_3005E10.unk_14);
+            if (gUnknown_3005E10.unk_30)
+            {
+                gUnknown_3005E10.unk_30 &= ~gUnknown_3005E10.unk_14;
+                for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu.c]: `RFU_CHILD_MAX`

File: src/link_rfu.c [hunk 2]
```diff
@@ -0,0 +1,1382 @@
...
+    gUnknown_3005E10.unk_0e = unk_0e_bak;
+}
+
+static void sub_80FE83C(u8 a0)
+{
+    u8 i;
+
+    if (gRfuLinkStatus->sendSlotNIFlag)
+    {
+        for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu.c]: `RFU_CHILD_MAX`

File: src/link_rfu.c [hunk 3]
```diff
@@ -0,0 +1,1382 @@
...
+        {
+            if (gRfuSlotStatusNI[i]->send.state & 0x8000 && gRfuSlotStatusNI[i]->send.bmSlot & a0)
+            {
+                rfu_changeSendTarget(0x20, i, gRfuSlotStatusNI[i]->send.bmSlot & ~a0);
+            }
+        }
+    }
+    if (gRfuLinkStatus->recvSlotNIFlag)
+    {
+        for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu.c]: `RFU_CHILD_MAX`

File: src/link_rfu.c [hunk 4]
```diff
@@ -0,0 +1,1382 @@
...
+{
+    u8 i;
+    u8 j;
+    u8 flags;
+
+    if (gUnknown_3005E10.unk_18)
+    {
+        if (gRfuLinkStatus->sendSlotNIFlag)
+        {
+            for (i = 0; i < 4; i ++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu.c]: `RFU_CHILD_MAX`

Comment (jiangzhengwenjz) [inline src/link_rfu.c]: no space before `++`

File: src/link_rfu.c [hunk 5]
```diff
@@ -0,0 +1,1382 @@
...
+        gUnknown_3005E10.unk_1a = 40;
+        gUnknown_3005E10.unk_07 = 4;
+        gUnknown_3005E10.unk_04 = 9;
+    }
+}
+
+static void sub_80FDC98(u16 r8, u16 r6)
+{
+    u8 sp0;
+    register u8 *stwiRecvBuffer asm("r0");
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/link_rfu.c]: Required to match in emerald

File: src/librfu_rfu.c [hunk 6]
```diff
@@ -1533,7 +1533,7 @@ void rfu_UNI_clearRecvNewDataFlag(u8 slotStatusIndex)
         gRfuSlotStatusUNI[slotStatusIndex]->recv.newDataFlag = 0;
 }
 
-void rfu_REQ_sendData(u8 clockChangeFlag)
+void rfu_REQ_sendData(bool8 clockChangeFlag)
 {
     if (gRfuLinkStatus->parentChild != MODE_NEUTRAL)
     {
```
Comment (jiangzhengwenjz) [inline src/librfu_rfu.c]: Can you fix line 1540 (`MODE_PARENT`)

Comment (jiangzhengwenjz) [reply src/librfu_rfu.c]: use `TRUE`/`FALSE` macro in this function for `clockChangeFlag`

File: src/link_rfu_2.c [hunk 7]
```diff
@@ -0,0 +1,2835 @@
...
+        break;
+    }
+    }
+}
+
+static void sub_80F8AA4(void)
+{
+    u8 i;
+    u8 r5 = gUnknown_3005E10.unk_00;
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 8]
```diff
@@ -0,0 +1,2835 @@
...
+            DestroyTask(taskId);
+            break;
+        }
+        break;
+    case 12:
+    {
+        u8 r5 = 1 << Rfu.unk_c3e;
+        rfu_clearSlot(12, Rfu.unk_c3e);
+        rfu_setRecvBuffer(16, Rfu.unk_c3e, Rfu.unk_c3f, 70);
+        rfu_UNI_setSendData(r5, Rfu.unk_4c, 14);
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `sizeof(Rfu.unk_4c)`

File: src/link_rfu_2.c [hunk 9]
```diff
@@ -0,0 +1,2835 @@
...
+            Rfu.unk_ce4 = 2;
+            DestroyTask(taskId);
+            break;
+        }
+        break;
+    case 12:
+    {
+        u8 r5 = 1 << Rfu.unk_c3e;
+        rfu_clearSlot(12, Rfu.unk_c3e);
+        rfu_setRecvBuffer(16, Rfu.unk_c3e, Rfu.unk_c3f, 70);
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `sizeof(Rfu.unk_c3f)`

File: src/link_rfu_2.c [hunk 10]
```diff
@@ -0,0 +1,2835 @@
...
+        break;
+    case 17:
+        sub_80FD538(2, 0, 240, gUnknown_843EC8C);
+        sub_80FEA10(sub_80F8D20);
+        Rfu.unk_04 = 18;
+        break;
+    case 18:
+        break;
+    case 13:
+        if (rfu_UNI_setSendData(1 << Rfu.unk_c3e, Rfu.unk_4c, 14) == 0)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `sizeof(Rfu.unk_4c)`

File: src/link_rfu_2.c [hunk 11]
```diff
@@ -0,0 +1,2835 @@
...
+            rfu_clearSlot(3, i);
+        }
+        r5 >>= 1;
+    }
+}
+
+static void sub_80F8AEC(void)
+{
+    u8 r5 = gUnknown_3005E10.unk_00;
+    rfu_UNI_setSendData(r5, Rfu.unk_c87, 70);
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `sizeof(Rfu.unk_c87)`

File: src/link_rfu_2.c [hunk 12]
```diff
@@ -0,0 +1,2835 @@
...
+                flags >>= 1;
+            }
+            sub_80F9038();
+            sub_80F9868(0);
+            CallRfuFunc();
+            if (Rfu.unk_ce5 && !Rfu.unk_cd9)
+            {
+                gUnknown_203AC08.unk_0e = 0;
+                rfu_clearSlot(3, Rfu.unk_cda);
+                for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 13]
```diff
@@ -0,0 +1,2835 @@
...
+                {
+                    if ((Rfu.unk_ce5 >> i) & 1)
+                    {
+                        rfu_setRecvBuffer(0x10, i, Rfu.unk_14[i], 14);
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `sizeof(Rfu.unk_14[i])`

File: src/link_rfu_2.c [hunk 14]
```diff
@@ -0,0 +1,2835 @@
...
+                    }
+                }
+                sub_80F887C(Rfu.unk_ce2, Rfu.unk_ce2 | Rfu.unk_ce5);
+                Rfu.unk_ce9 = Rfu.unk_ce5;
+                Rfu.unk_ce2 |= Rfu.unk_ce5;
+                Rfu.unk_ce5 = 0;
+                rfu_UNI_setSendData(Rfu.unk_ce2, Rfu.unk_c87, 70);
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: sizeof(Rfu.unk_c87)

File: src/link_rfu_2.c [hunk 15]
```diff
@@ -0,0 +1,2835 @@
...
+    Rfu.unk_80[a0].unk_12 = 0;
+}
+
+static u8 sub_80F9770(const u8 *a0)
+{
+    u8 i;
+
+    if (Rfu.unk_0c == 1)
+        return FALSE;
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 16]
```diff
@@ -0,0 +1,2835 @@
...
+        if (sub_80F9C50())
+            gSendCmd[1] = Rfu.unk_5a;
+        break;
+    case 0x7700:
+    case 0x7800:
+        tmp = Rfu.unk_ce2 ^ Rfu.unk_ce3;
+        Rfu.playerCount = gUnknown_843EC41[tmp] + 1;
+        gSendCmd[1] = Rfu.playerCount;
+        buff = (u8 *)(gSendCmd + 2);
+        for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 17]
```diff
@@ -0,0 +1,2835 @@
...
+    {
+        if (gUnknown_843EC8C[i] == 0xFFFF)
+            return FALSE;
+    }
+    return TRUE;
+}
+
+u8 sub_80FA484(bool32 a0)
+{
+    if (a0 == FALSE)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `!a0`

File: src/link_rfu_2.c [hunk 18]
```diff
@@ -0,0 +1,2835 @@
...
+    return 0;
+}
+
+void sub_80FA4A8(void)
+{
+    Rfu.unk_cd9 = 1;
+    sub_80FD760(FALSE);
+}
+
+u8 rfu_get_multiplayer_id(void)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `LinkRfu_GetMultiplayerId`, maybe

File: src/link_rfu_2.c [hunk 19]
```diff
@@ -0,0 +1,2835 @@
...
+}
+
+u8 GetRfuPlayerCount(void)
+{
+    return Rfu.playerCount;
+}
+
+bool8 IsLinkRfuTaskFinished(void)
+{
+    return Rfu.RfuFunc ? FALSE : TRUE;
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `!= NULL`

File: src/link_rfu_2.c [hunk 20]
```diff
@@ -0,0 +1,2835 @@
...
+}
+
+static void CallRfuFunc(void)
+{
+    if (Rfu.RfuFunc)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `!= NULL`

File: src/link_rfu_2.c [hunk 21]
```diff
@@ -0,0 +1,2835 @@
...
+        }
+    }
+    return retval;
+}
+
+bool32 sub_80FA5D4(void)
+{
+    u8 flags = 0;
+    s32 i;
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 22]
```diff
@@ -0,0 +1,2835 @@
...
+            flags |= (1 << i);
+            Rfu.unk_cd5[i] = 0;
+        }
+    }
+    if (flags)
+    {
+        rfu_REQ_disconnect(flags);
+        rfu_waitREQComplete();
+    }
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 23]
```diff
@@ -0,0 +1,2835 @@
...
+    u32 r5;
+    u32 r7;
+    s32 r8;
+
+    if (sub_80F9800()->unk_0a_0 == 0x45)
+    {
+        r5 = 0;
+        r7 = 0;
+        r8 = Rfu.unk_ce2 ^ Rfu.unk_ce3;
+        for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 24]
```diff
@@ -0,0 +1,2835 @@
...
+    ResetTasks();
+    ResetPaletteFade();
+    SetVBlankCallback(sub_80FBA64);
+    if (IsWirelessAdapterConnected())
+    {
+        gLinkType = LINKTYPE_0x1111;
+        sub_800B1F4();
+        OpenLink();
+        SeedRng(gMain.vblankCounter2);
+        for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `NELEMS(gSaveBlock2Ptr->playerTrainerId)`

File: src/link_rfu_2.c [hunk 25]
```diff
@@ -0,0 +1,2835 @@
...
+}
+
+void sub_80FBD6C(u32 a0)
+{
+    if (a0 != 0)
+    {
+        s32 i;
+        u8 var = 0;
+
+        for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 26]
```diff
@@ -0,0 +1,2835 @@
...
+    '\0'
+};
+
+static const char gUnknown_843EE57[9] = {
+    ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ',
+    '\0'
+};
+
+static const char gUnknown_843EE60[] = {' ', '\0'};
+static const char gUnknown_843EE62[] = {'*', '\0'};
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: are these C strings?

File: src/link_rfu_2.c [hunk 27]
```diff
@@ -0,0 +1,2835 @@
...
+static void sub_80FC228(void)
+{
+    s32 i, j;
+
+    nullsub_88(GetBlockReceivedStatus(), 0x1C, 0x13, 2);
+    nullsub_88(gRfuLinkStatus->connSlotFlag, 0x14, 1, 1);
+    nullsub_88(gRfuLinkStatus->linkLossSlotFlag, 0x17, 1, 1);
+    if (Rfu.unk_0c == 1)
+    {
+        for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 28]
```diff
@@ -0,0 +1,2835 @@
...
+
+static void sub_80F8AA4(void)
+{
+    u8 i;
+    u8 r5 = gUnknown_3005E10.unk_00;
+    for (i = 0; i < 4; i++)
+    {
+        if (r5 & 1)
+        {
+            rfu_setRecvBuffer(16, i, Rfu.unk_14[i], 14);
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `sizeof(Rfu.unk_14[i])`

File: src/link_rfu_2.c [hunk 29]
```diff
@@ -0,0 +1,2835 @@
...
+        return 1;
+    return 0;
+}
+
+static void sub_80FA738(void)
+{
+    s32 i;
+
+    sub_80FA528();
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 30]
```diff
@@ -0,0 +1,2835 @@
...
+        gTasks[taskId].data[0]++;
+        break;
+    case 6:
+        DestroyTask(taskId);
+        gReceivedRemoteLinkPlayers = 1;
+        Rfu.unk_ce8 = 0;
+        sub_80FEA34(1, 0x258);
+        if (Rfu.unk_ce6)
+        {
+            for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX` (determined from another function)

File: src/link_rfu_2.c [hunk 31]
```diff
@@ -0,0 +1,2835 @@
...
+        }
+        break;
+    }
+}
+
+static void sub_80FA9D0(u16 a0)
+{
+    s32 i;
+
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 32]
```diff
@@ -0,0 +1,2835 @@
...
+        if ((a0 >> i) & 1)
+            Rfu.unk_cde[i] = 0;
+    }
+}
+
+static void sub_80FA9FC(const struct UnkRfuStruct_8010A14 *a0)
+{
+    s32 i;
+    Rfu.playerCount = a0->unk_0f;
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 33]
```diff
@@ -0,0 +1,2835 @@
...
+            gLinkPlayers[r4] = r2->linkPlayer;
+            sub_800B284(gLinkPlayers + r4);
+            gTasks[taskId].data[0]++;
+        }
+        break;
+    case 3:
+        r5 = (struct UnkRfuStruct_8010A14 *)gBlockSendBuffer;
+        memcpy(r5->unk_00, "PokemonSioInfo", sizeof("PokemonSioInfo"));
+        r5->unk_0f = Rfu.playerCount;
+        for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 34]
```diff
@@ -0,0 +1,2835 @@
...
+        break;
+    case 5:
+        if (IsLinkTaskFinished() && GetBlockReceivedStatus() & 1)
+        {
+            CpuFill16(0, gBlockRecvBuffer, sizeof(struct UnkRfuStruct_8010A14));
+            ResetBlockReceivedFlag(0);
+            Rfu.unk_ce8 = 0;
+            if (Rfu.unk_ce6)
+            {
+                for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 35]
```diff
@@ -0,0 +1,2835 @@
...
+    switch (a0)
+    {
+    case 0x00:
+        Rfu.unk_04 = 2;
+        break;
+    case 0x10:
+        break;
+    case 0x11:
+        sub_80FB564(gUnknown_3005E10.unk_14);
+        for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 36]
```diff
@@ -0,0 +1,2835 @@
...
+        Rfu.unk_cdb = 1;
+        break;
+    }
+}
+
+static void sub_80FB564(s32 a0)
+{
+    s32 i;
+
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 37]
```diff
@@ -0,0 +1,2835 @@
...
+        }
+    }
+}
+
+static u8 sub_80FB5A0(s32 a0)
+{
+    u8 ret = 0;
+    u8 i;
+
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 38]
```diff
@@ -0,0 +1,2835 @@
...
+    const u8 *ptr_ = ptr;
+    return (ptr_[1] << 8) | (ptr_[0]);
+}
+
+static u8 sub_80FBC70(const u8 *a0, u16 a1)
+{
+    u8 i;
+    u8 ret = 0xFF;
+
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 39]
```diff
@@ -0,0 +1,2835 @@
...
+        return TRUE;
+    else
+        return FALSE;
+}
+
+bool32 sub_80FC1CC(void)
+{
+    s32 i;
+
+    for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: src/link_rfu_2.c [hunk 40]
```diff
@@ -0,0 +1,2835 @@
...
+    if (Rfu.unk_0c == 1)
+    {
+        for (i = 0; i < 4; i++)
+        {
+            if ((gRfuLinkStatus->getNameFlag >> i) & 1)
+            {
+                nullsub_88(gRfuLinkStatus->partner[i].serialNo, 1, i + 3, 4);
+                nullsub_87((void*) &gRfuLinkStatus->partner[i].gname, 6, i + 3);
+                nullsub_87(gRfuLinkStatus->partner[i].uname, 0x16, i + 3);
+
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: `RFU_CHILD_MAX`

File: include/link_rfu.h [hunk 41]
```diff
@@ -4,14 +4,220 @@
 #include "global.h"
 #include "librfu.h"
 
+// RfuTgtData.gname is read as these structs.
+struct GFtgtGnameSub
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply include/link_rfu.h]: Nah.

File: src/link_rfu_2.c (hunk 26 above)
Comment (PikalaxALT) [reply src/link_rfu_2.c]: If they were, they'd be word aligned.

File: src/link_rfu_2.c [hunk 42]
```diff
@@ -0,0 +1,2834 @@
...
+        }
+        for (r4 = 4; r4 != 0; r4--)
+        {
+            for (i = 0; i < RFU_CHILD_MAX && Rfu.unk_cde[i] != r4; i++);
+            if (i == 4)
+            {
+                r6 = r4;
+            }
+        }
+        for (r5 &= ~r2, i = 0; i < 4; r5 >>= 1, i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: You missed this one

File: src/link_rfu_2.c [hunk 43]
```diff
@@ -0,0 +1,2834 @@
...
+}
+
+void sub_80FBD6C(u32 a0)
+{
+    if (a0 != 0)
+    {
+        s32 i;
+        u8 var = 0;
+
+        for (i = 0; i < 4; i++)
```
Comment (jiangzhengwenjz) [inline src/link_rfu_2.c]: You missed this one.

**Pull request merged**